### PR TITLE
methoderror for constructors considers typevars

### LINKED ISSUE #33539: Methoderror for constructors should consider typevars
Consider
```
julia> struct Foo{T,S}
           a::S
           b::T
       end

julia> Foo(1, 2)
Foo{Int64,Int64}(1, 2)

julia> Foo{Int64, Int64}(1, 2)
Foo{Int64,Int64}(1, 2)

julia> Foo{Int64}(1)
ERROR: MethodError: no method matching Foo{Int64,S} where S(::Int64)
Stacktrace:
 [1] top-level scope at REPL[4]:1
```
The `MethodError` here is missing closest candidates for the implicitly defined constructors.
The reason for that is likely the following:
```
julia> methods(Foo{Int64})
# 0 methods for type constructor:

julia> methods(Foo)
# 1 method for type constructor:
[1] (::Type{Foo})(a::S, b::T) where {T, S} in Main at REPL[1]:2

julia> methods(Base.unwrap_unionall(Foo))
# 2 methods for type constructor:
[1] (::Type{Foo})(a::S, b::T) where {T, S} in Main at REPL[1]:2
[2] (::Type{Foo{T,S}})(a, b) where {T, S} in Main at REPL[1]:
```

so one approach to start here might be widening `f` 
 in https://github.com/JuliaLang/julia/blob/master/base/errorshow.jl#L350 to `Base.unwrap_unionall(Base.unwrap_unionall(Foo{Int64}).name.wrapper)` or something similar.

### PR DESCRIPTION
**Pull request by KlausC**
fix #33539

### CODE REVIEW

File: base/errorshow.jl [hunk 1]
```diff
@@ -359,8 +359,10 @@ function show_method_candidates(io::IO, ex::MethodError, @nospecialize kwargs=()
         end
     end
 
-    for (func, arg_types_param) in funcs
-        for method in methods(func)
+    for (f, arg_types_param) in funcs
+        f2 = f isa UnionAll ? unwrap_unionall(unwrap_unionall(f).name.wrapper) : f
```
Comment (vtjnash) [inline base/errorshow.jl]: `unwrap_unionall` doesn't necessarily have a field `name` (if it's not a DataType). Most nearly, you can get `typename(f)`, although that also may throw errors

Comment (KlausC) [reply base/errorshow.jl]: I will check for `unwrap_unionall(f) isa DataType`.

Only to be curious: which cases are there, where `f` is a `UnionAll` and `unwrap_unionall(f)` is not a `DataType`?

File: base/errorshow.jl [hunk 2]
```diff
@@ -372,16 +374,8 @@ function show_method_candidates(io::IO, ex::MethodError, @nospecialize kwargs=()
...
-            if !isa(func, rewrap_unionall(s1, method.sig))
-                # function itself doesn't match
-                continue
-            else
-                # TODO: use the methodshow logic here
-                use_constructor_syntax = isa(func, Type)
-                print(iob, use_constructor_syntax ? func : typeof(func).name.mt.name)
-            end
-            print(iob, "(")
+            print(iob, "  ", func, "(")
```
Comment (vtjnash) [inline base/errorshow.jl]: Something like this to still highlight mismatches, like other arguments?
```suggestion
print(iob, "  ")
if !isa(func, rewrap_unionall(s1, method.sig))
    ... Base.with_output_color(...); print(iob, func); end ...
else
    print(iob, func)
end
print(iob, "("))
```

File: base/errorshow.jl [hunk 3]
```diff
@@ -359,8 +359,10 @@ function show_method_candidates(io::IO, ex::MethodError, @nospecialize kwargs=()
         end
     end
 
-    for (func, arg_types_param) in funcs
-        for method in methods(func)
+    for (f, arg_types_param) in funcs
+        f2 = f isa UnionAll ? unwrap_unionall(unwrap_unionall(f).name.wrapper) : f
+        for method in methods(f2)
```
Comment (vtjnash) [inline base/errorshow.jl]: `methods(unwrap_unionall(...))` is an invalid query—this must be called only on a properly wrapped object (e.g. the `.wrapper`)

Comment (KlausC) [reply base/errorshow.jl]: That was the original suggestion in https://github.com/JuliaLang/julia/issues/33539#issue-506110561, and it works obviously for `DataType`.
Using the `methods(... .name.wrapper)` does not deliver all the methods we are after.
How would I have to "properly wrap" the object?

File: base/errorshow.jl (hunk 2 above)
Comment (KlausC) [reply base/errorshow.jl]: IMO that will be a very bad effort/result ratio, it the `...` are extended. I would prefer to leave it as is.

**Comment by KlausC** [comment]
Bump.

**Comment by StefanKarpinski** [comment]
@vtjnash has been traveling but it would be good to make forward progress on this.

File: base/errorshow.jl (hunk 3 above)
Comment (vtjnash) [reply base/errorshow.jl]: Ah, right. `methods` can't do this query properly (which is its own shortcoming). You'd need to bypass it and call into the `Base._methods_by_ftype` function directly with `f2 isa Type ? Type{<:f2} : typeof(f2)`.

Comment (KlausC) [reply base/errorshow.jl]: I don't see the point: If I `@enter methods(f2)` I end at
```
830      tt = signature_type(f, t)
831      return _methods_by_ftype(tt, lim, world)

1|julia> tt
Tuple{Type{Foo{T<:Integer,S,U}},Vararg{Any,N} where N}
```
That seems to be exactly the same, what you propose to call directly. The results are identical. Only difference is,
the `methods` throws an `ArgumentError`, if `f2 isa Core.Builtin`. So why shouldn't I simply call `methods(f2)`??

**Comment by KlausC** [comment]
there has been progress with the changed files, which resolve the issue better than this PR.

**Pull request closed without merging**